### ISSUE TITLE
Option to create data with a hash to avoid duplicate imports

### ISSUE THREAD

**Issue by gmaclennan** [digidem/mapeo-core-next#507]
## Description

Normally we use a random ID for new records. This can lead to duplicates when importing data: the same data can be imported twice, resulting in duplicate records with different IDs but the same information.

For some record types, having duplicate records with the same information is ok, e.g. for observations we could have two observations created at different times with the same `value`, and they are intentionally different. However for some record types, particularly related to imports, two records with the same value are an actual duplicate.

To avoid this we could add an option to use a hash of the value as the ID, and check for duplicates before creation, e.g.

```js
import stableStringify from 'json-stable-stringify'
import { createHash } from 'crypto'

function hashObject(obj) {
  return createHash('sha256').update(stableStringify(obj)).digest()
}

class DataType {
  // ...
  async create(value, { useHashAsId }) {
    let docId
    if (useHashAsId) {
      docId = hashObject(value).toString('hex')
      const existing = await this.getByDocId(docId).catch(() => null)
      if (existing) return // or throw?
    } else {
      docId = randomBytes(32).toString('hex')
    }
    // ...
  }
}
```

Not sure if this is best as an option for create, or we just keep it internal as a helper for import functions (like config import) and use the internal `createWithDocId` method.

This would be useful for config imports to avoid adding a deletion entry to then replace it with the exact same value.

**Comment by EvanHahn** [comment]
I think the questions around data de-duplication are difficult to generalize. Before devising a generic solution, I would address this just for config imports. I suspect an elegant generic solution will be best informed by concrete cases such as this.

Regardless of whether we do a generic or targeted solution, I can think of a few questions around de-duping:

- How should race conditions be handled? Imagine that two processes insert a record this way. Both might check for duplicates; seeing none, they _both_ create a record.
- How should duplicates be handled from other devices? There's nothing to stop a buggy peer from sending you a duplicate record.
- How should hashing work? Is `json-stable-stringify` sufficient for all the data we want to store? Are `{ foo: undefined }` and `{}` equivalent? How do we _change_ a record's hash function if user requirements change?
- How does this impact performance?

I think these questions are easier to answer on a case-by-case basis (i.e., just for config import). And once we have a few answers, we might see some patterns.

**Comment by gmaclennan** [comment]
> I think the questions around data de-duplication are difficult to generalize. Before devising a generic solution, I would address this just for config imports. I suspect an elegant generic solution will be best informed by concrete cases such as this.

Agreed. I was thinking about this in the context of importing alerts (e.g. deforestation alerts) and having a way to avoid duplicates, but I think focussing on config imports is helpful for now.

> Regardless of whether we do a generic or targeted solution, I can think of a few questions around de-duping:
> 
> * How should race conditions be handled? Imagine that two processes insert a record this way. Both might check for duplicates; seeing none, they _both_ create a record.
> * How should duplicates be handled from other devices? There's nothing to stop a buggy peer from sending you a duplicate record.

I think both of these should be handled by our existing `getWinner` strategy. Two records with the same ID would be treated as a fork (although we do not currently have a test for a "fork" when forked records do not have any `links` - the code was written on the assumption that only one record has no links, but I think it does not depend on that. The `getWinner` will choose the most recently added to the database, and if dates are the same then will deterministically choose one based on versionId. A buggy peer that incorrectly hashes to match an ID while the data is different would cause issues, but I think that is a larger issue beyond just hashing and deduplication.

> * How should hashing work? Is `json-stable-stringify` sufficient for all the data we want to store? Are `{ foo: undefined }` and `{}` equivalent? How do we _change_ a record's hash function if user requirements change?

I think in terms of how we currently store data as protobufs, there is no distinction between these two options, so I think stable JSON stringify should be sufficient.

> * How does this impact performance?
> 
> I think these questions are easier to answer on a case-by-case basis (i.e., just for config import). And once we have a few answers, we might see some patterns.

**Comment by EvanHahn** [comment]
How would you feel about either (1) closing this issue (2) changing it to be about imports specifically?

**Comment by gmaclennan** [comment]
Agreed about closing this and addressing this in imports internally by leveraging the `createWithDocId` code, rather than adding something to the public API.

**Comment by EvanHahn** [comment]
Closing in favor of the (more specific) #513.
